File: python/46 - Permutations.py
# 46
# Permutations
# https://leetcode.com/problems/permutations/
# python
# medium
# O(n*n!)
# O(1)
# permute

# solution using running list of options and path
# time: O(n*n!)
# space: O(n^2)
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        results = []

        def recurse(path, options):
            if len(options) == 1:
                results.append(path[:] + options[:])
            else:
                for i in range(len(options)):
                    recurse(path + [options[i]], options[:i] + options[i + 1 :])

        recurse([], nums)
        return results


# initial swapping solution
# time: O(n*n!)
# space: O(1) excluding output space
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        results = []
        end = len(nums)

        def recurse(start):
            if start == end:
                results.append(nums[:])
            for i in range(start, end):
                nums[start], nums[i] = nums[i], nums[start]
                recurse(start + 1)
                nums[start], nums[i] = nums[i], nums[start]

        recurse(0)
        return results


# another recursive algorithm
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []

        def visit(curr, pool):
            if not pool:
                ans.append(curr)
            for i in range(len(pool)):
                visit(curr + [pool[i]], pool[:i] + pool[i + 1 :])

        visit([], nums)
        return ans

File: c++/46 - Permutations.cpp
// 46
// Permutations
// https://leetcode.com/problems/permutations/
// c++
// medium
// O(n*n!)
// O(1)
// permute

// memory effecient swapping algorthim
// time: O(n*n!) (multiplied by n due to copying of vector into result which takes linear time)
// space: O(1)  excluding output space ofc
class Solution {
public:
    void visit(int begin, vector<int>& nums, vector<vector<int>>& results) {
        if (begin == nums.size())
            results.push_back(nums);
        for (int i = begin; i < nums.size(); i++) {
            swap(nums[i], nums[begin]);
            visit(begin+1, nums, results);
            swap(nums[i], nums[begin]);
        }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> results;
        visit(0, nums, results);
        return results;
    }
};